### ISSUE TITLE
[BUG] Reverse-mode AD produces incorrect gradients when arrays are reused across or within tape operations

### ISSUE THREAD

**Issue by mar-yan24** [NVIDIA/warp#1294]
## Bug Description

Hi warp team. I hope the bugs aren't too overwhelming lol but I have another issue with tape for autodiff. Actually, this bug is technically more like two.

### Bug A: Self-aliasing, kernel reads and writes the same array

Basically what is happening is that I am trying to do a pretty simple operation of something along the lines of `arr[i] = arr[i] + 1`, where the kernel reads and writes the same array. I think what is happening is that Warp's backward pass is treating `arr` as both input and output. Here is the MRE:

```
import warp as wp
import numpy as np

wp.init()


@wp.kernel
def sum_kernel(data: wp.array(dtype=float), loss: wp.array(dtype=float)):
    wp.atomic_add(loss, 0, data[wp.tid()])


@wp.kernel
def add_one(arr: wp.array(dtype=float)):
    i = wp.tid()
    arr[i] = arr[i] + 1.0  # reads and writes same array


src = wp.array([1.0, 2.0, 3.0], dtype=float, requires_grad=True)
arr = wp.zeros(3, dtype=float, requires_grad=True)
loss = wp.zeros(1, dtype=float, requires_grad=True)

tape = wp.Tape()
with tape:
    wp.copy(arr, src)
    wp.launch(add_one, dim=3, inputs=[arr])
    wp.launch(sum_kernel, dim=3, inputs=[arr, loss])
tape.backward(loss)

grad = src.grad.numpy()
expected = np.array([1.0, 1.0, 1.0])
print(f"got: {grad}")
print(f"expected: {expected}")
```

### Bug B: Gradient leak through overwritten values

For this one, when two operations write to the same output array, and the second one overwrites the first, the backward pass still propagates gradient through the first operation. When running this, I expected/want those values to be thrown away and shouldnt affect the loss. Here is the MRE:

```
import warp as wp
import numpy as np

wp.init()


@wp.kernel
def sum_kernel(data: wp.array(dtype=float), loss: wp.array(dtype=float)):
    wp.atomic_add(loss, 0, data[wp.tid()])


@wp.kernel
def scale(inp: wp.array(dtype=float), out: wp.array(dtype=float), s: float):
    i = wp.tid()
    out[i] = inp[i] * s


inp = wp.array([1.0, 1.0, 1.0], dtype=float, requires_grad=True)
out = wp.zeros(3, dtype=float, requires_grad=True)
loss = wp.zeros(1, dtype=float, requires_grad=True)

tape = wp.Tape()
with tape:
    wp.launch(scale, dim=3, inputs=[inp, out, 2.0])  # out = inp*2
    wp.launch(scale, dim=3, inputs=[inp, out, 3.0])  # out = inp*3 (overwrites)
    wp.launch(sum_kernel, dim=3, inputs=[out, loss])
tape.backward(loss)

grad = inp.grad.numpy()
expected = np.array([3.0, 3.0, 3.0])
print(f"got: {grad}")
print(f"expected: {expected}")
```

For both of these issues, disjoint slice writes to the same array produce correct gradients. Pretty sure these bugs are scoped to the overlapping and full overwrites. I also am confident that these bugs stem from the same issue which is that tape is not tracking array mutations at the element level. 

Let me know if any more information is needed and thank you Warp team for the help!

### System Information

Warp 1.12.0
CUDA 12.9
Windows 10, NVIDIA GTX 1070
Python 3.13

**Comment by shi-eric** [comment]
Hey @mar-yan24, are you familiar with the [differentiability](https://nvidia.github.io/warp/user_guide/differentiability.html) section in the Warp docs? I think the docs address the question asked here and also the one in the previous issue you filed, but please let us know if anything continues to be unclear after reading that page.

**Comment by mar-yan24** [comment]
@shi-eric ah I see, sorry I should have read the dynamics loop section closer. This makes a lot more sense with both issues lol. Thank you again for the help, sorry I should've looked closer at the documentation.

**Comment by shi-eric** [comment]
No worries @mar-yan24! Glad you were able to figure things out.
